tests/web/QueueManager.cpp

```diff
@@ -0,0 +1,169 @@
+  };
+
+  std::function<int()> get_epochs = [](){return 50;};
+  std::function<std::string()> metric_fun = [](){return "Hello";};
+
+  emp::SettingConfig config;
+  emp::Ptr<emp::QueueManager> queue_manager;
+
+  Test_QueueManager()
+  : BaseTest({"emp_test_container"}) // we can tell BaseTest that we want to create a set of emp::web::Document
+                                     // objects for each given html element ids.
+  {
+    
+    config.AddSetting<int>("my_param") = {6};
+    queue_manager.New(config);
+
+    queue_manager->AddMetric(metric_fun, "Hello metric");
+
+    queue_manager->AddQueueButton(get_setting_config, get_epochs); // Add button and text input
+    queue_manager->BuildTable("my_table"); // Add progress table
+    
+    Doc("emp_test_container")
+      << queue_manager->GetDiv();
+
+    // Test adding metric after table built
+    queue_manager->AddMetric(metric_fun, "Late Hello");
+
+    queue_manager->AddRun(get_setting_config(), get_epochs());
+    queue_manager->AddNewQueuedRunToTable();
+
+    emp_assert(!queue_manager->IsEmpty());
+    emp_assert(queue_manager->RunsRemaining() == 1);
+    emp_assert(queue_manager->FrontRun().GetEpoch() == 0);        
+    emp_assert(queue_manager->FrontRun().GetConfig().GetValue<int>("my_param") == 9);
+    emp_assert(queue_manager->FrontRun().GetNEpochs() == 50);
+
+    queue_manager->FrontRun().IncEpoch();
+    queue_manager->Update();
+
+    queue_manager->AddRun(config, 20);
+    queue_manager->AddNewQueuedRunToTable();
+
+    emp_assert(queue_manager->RunsRemaining() == 2);
+    queue_manager->FrontRun().IncEpoch(49);
+    queue_manager->Update();
+    
+    // Previous run should have finished
+    emp_assert(queue_manager->RunsRemaining() == 1);
+  }
+
+  ~Test_QueueManager() {
+    queue_manager->ResetDiv();
+    queue_manager.Delete();
+  }
+
+  void Describe() override {
+
+    EM_ASM({
+      describe("emp::QueueManager GUI", function() {
+
+        describe("data table", function() {
+
+          it('should exist and be a table', function() {
+            chai.assert.equal($( "table#my_table" ).length, 1);
+          });
+
+          it('should have grandparent #emp_test_container', function() {
+            const grand_parent_id = $("#my_table").parent().parent().attr("id");
+            chai.assert.equal(grand_parent_id, "emp_test_container");
+          });
+
+          it('should have 4 columns', function() {
+            columns = $("#my_table").find("th>span");
+            chai.assert.equal(columns.length, 5);
+            chai.assert.equal(columns[0].firstChild.textContent, "Run");
+            chai.assert.equal(columns[1].firstChild.textContent, "my_param");
+            chai.assert.equal(columns[2].firstChild.textContent, "Epoch");
+            chai.assert.equal(columns[3].firstChild.textContent, "Hello metric");
+            chai.assert.equal(columns[4].firstChild.textContent, "Late Hello");
+          });
+
+          it('should have 3 rows', function() {
```

Comment:
love chai's descriptive syntax here

Reply:
It's so nice. Almost makes me wish I had more things to test in Javascript :)